### PR TITLE
Enable recursive type definitions

### PR DESCRIPTION
**Pull request by yilinjuang**
Mypy supports recursive types now.

https://github.com/python/mypy/issues/731
https://github.com/python/mypy/issues/14219

### CODE REVIEW

File: src/graphql/pyutils/path.py [hunk 1]
```diff
@@ -2,15 +2,15 @@
...
+from typing import Any, NamedTuple, Optional
 
 __all__ = ["Path"]
 
 
 class Path(NamedTuple):
     """A generic path of string or integer indices"""
 
-    prev: Any  # Optional['Path'] (python/mypy/issues/731)
+    prev: Optional[Path]
```
Comment (Cito) [inline src/graphql/pyutils/path.py]:
```suggestion
    prev: Path |  None
```

recently switched to this newer style

File: src/graphql/validation/rules/overlapping_fields_can_be_merged.py [hunk 2]
```diff
@@ -98,10 +98,7 @@ def enter_selection_set(self, selection_set: SelectionSetNode, *_args: Any) -> N
 # Field name and reason.
 ConflictReason: TypeAlias = Tuple[str, "ConflictReasonMessage"]
 # Reason is a string, or a nested list of conflicts.
-if MYPY:  # recursive types not fully supported yet (/python/mypy/issues/731)
```
Comment (Cito) [inline src/graphql/validation/rules/overlapping_fields_can_be_merged.py:101]: You can now remove the MYPY constant completely.

File: src/graphql/pyutils/path.py (hunk 1 above)
Comment (Cito) [reply src/graphql/pyutils/path.py]: Line 28 must then also be changed to `curr: Path | None = self`.

Also, in `test_executor` we must now add the line `assert prev is not None` two times (as lines 311, 315).

**Review by Cito — APPROVED** [review]
Thanks you for contributing.

There are still some minor issues, see failing QA test and comments.

File: src/graphql/validation/rules/overlapping_fields_can_be_merged.py (hunk 2 above)
Comment (yilinjuang) [reply src/graphql/validation/rules/overlapping_fields_can_be_merged.py:101]: Removed

File: src/graphql/pyutils/path.py (hunk 1 above)
Comment (yilinjuang) [reply src/graphql/pyutils/path.py]: Fixed

**Pull request merged**